pywinauto/application.py

```diff
@@ -1,1324 +1,1331 @@
+        return ctrls[-1]
+
+    def child_window(self, **criteria):
+        """
+        Add criteria for a control
+
+        When this window specification is resolved it will be used
+        to match against a control.
+        """
+        # default to non top level windows because we are usualy
+        # looking for a control
+        if 'top_level_only' not in criteria:
+            criteria['top_level_only'] = False
+
+        new_item = WindowSpecification(self.criteria[0])
+        new_item.criteria.extend(self.criteria[1:])
+        new_item.criteria.append(criteria)
+
+        return new_item
+
+    def window(self, **criteria):
+        """Deprecated alias of child_window()"""
+        warnings.warn(
+            "WindowSpecification.Window() WindowSpecification.Window_(), "
+            "WindowSpecification.window() and WindowSpecification.window_() "
+            "are deprecated, please switch to WindowSpecification.child_window()",
+            DeprecationWarning)
+        return self.child_window(**criteria)
+
+    def __getitem__(self, key):
+        """
+        Allow access to dialogs/controls through item access
+
+        This allows::
+
+            app['DialogTitle']['ControlTextClass']
+
+        to be used to access dialogs and controls.
+
+        Both this and :func:`__getattribute__` use the rules outlined in the
+        HowTo document.
+        """
+        # if we already have 2 levels of criteria (dlg, control)
+        # then resolve the control and do a getitem on it for the
+        if len(self.criteria) >= 2:
+
+            ctrls = self.__resolve_control(self.criteria)
+
+            # try to return a good error message if the control does not
+            # have a __getitem__() method)
+            if hasattr(ctrls[-1], '__getitem__'):
+                return ctrls[-1][key]
+            else:
+                message = "The control does not have a __getitem__ method " \
+                    "for item access (i.e. ctrl[key]) so maybe you have " \
+                    "requested this in error?"
+
+                raise AttributeError(message)
+
+        # if we get here then we must have only had one criteria so far
+        # so create a new :class:`WindowSpecification` for this control
+        new_item = WindowSpecification(self.criteria[0])
+
+        # add our new criteria
+        new_item.criteria.append({"best_match": key})
+
+        return new_item
+
+    def __getattribute__(self, attr_name):
+        """
+        Attribute access for this class
+
+        If we already have criteria for both dialog and control then
+        resolve the control and return the requested attribute.
+
+        If we have only criteria for the dialog but the attribute
+        requested is an attribute of DialogWrapper then resolve the
+        dialog and return the requested attribute.
+
+        Otherwise delegate functionality to :func:`__getitem__` - which
+        sets the appropria
```

Comment:
A good spot on the explicit call to ```self.__resolve_control``` and timings arguments! I have to admit, I don't understand why ```exists``` needs a special treatment, but if it's the case the ```exists``` processing should be moved a bit upper, before entering into the ```for``` loop, like this: 
```python 
if 'exists' in check_names:
    check = getattr(self, check_name)
    return check(retry_interval, float(retry_interval) // 2)

for check_name in check_names:
    try:
        ...
``` 
The current implementation will demonstrate unstable behavior if ```exists``` will be coupled with any other check.